fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/MultidimensionalIndexScanBounds.java

```diff
@@ -0,0 +1,314 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.annotation.API;
+import com.apple.foundationdb.async.RTree;
+import com.apple.foundationdb.record.EndpointType;
+import com.apple.foundationdb.record.IndexScanType;
+import com.apple.foundationdb.record.RecordCoreException;
+import com.apple.foundationdb.record.TupleRange;
+import com.apple.foundationdb.tuple.Tuple;
+import com.apple.foundationdb.tuple.TupleHelpers;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * {@link IndexScanBounds} for a multidimensional index scan. A multidimensional scan bounds object contains a
+ * {@link #prefixRange} and a {@link SpatialPredicate} which can be almost arbitrarily complex. The prefix range
+ * is a regular tuple range informing the index maintainer how to constrain the search over the non-multidimensional
+ * fields that can be viewed as a prefix whose data is stored in a regular one-dimensional index. The spatial predicate
+ * implements methods to quickly establish geometric overlap and containment. Spatial predicates do have internal
+ * structure as they can delegate to contained other spatial predicates thus allowing to form e.g. logical conjuncts
+ * or disjuncts.
+ */
+@API(API.Status.EXPERIMENTAL)
+public class MultidimensionalIndexScanBounds implements IndexScanBounds {
+    @Nonnull
+    private final TupleRange prefixRange;
+
+    @Nonnull
+    private final SpatialPredicate spatialPredicate;
+
+    public MultidimensionalIndexScanBounds(@Nonnull final TupleRange prefixRange, @Nonnull final SpatialPredicate spatialPredicate) {
+        this.prefixRange = prefixRange;
+        this.spatialPredicate = spatialPredicate;
+    }
+
+    @Nonnull
+    @Override
+    public IndexScanType getScanType() {
+        return IndexScanType.BY_VALUE;
+    }
+
+    @Nonnull
+    public TupleRange getPrefixRange() {
+        return prefixRange;
+    }
+
+    @Nonnull
+    public SpatialPredicate getSpatialPredicate() {
+        return spatialPredicate;
+    }
+
+    /**
+     * Method to compute if the rectangle handed in overlaps with this scan bounds object. This method is invoked when
+     * the R-tree data structure of a multidimensional index is searched. Note that this method can be implemented using
+     * a best-effort approach as it is permissible to indicate overlap between {@code mbr} and {@code this} when there
```

Comment:
Perhaps the method name should reflect this relaxed contract.

Reply:
`overlapsMbrApproximately`